### PR TITLE
exec: command fails: copy data out gracefully

### PR DESCRIPTION
**Pull request by TomasTomecek**
```
When the pod dies and exec is still happening, docker/openshift kill the
exec with rc=137. Copying data after this happens ends with:

https://prod.packit.dev/srpm-build/16278/logs

    WARNING: cannot use rsync: rsync not available in container

Would be nice if oc actually told us the pod is dead and not that
useless warning.

So instead of breaking there, we eat the exception and continue
execution to raise SandcastleCommandFailed which contains all the data
regarding the failure.

Those "weird" changes in api.py are done for sake of the new test case.
```

\+
```
increase pod timeout to 30 minutes

use case: it takes a long time to create an archive for RPM CI builds so
we need to wait for it before proceding with creating an SRPM
```

### CODE REVIEW

File: sandcastle/api.py [hunk 1]
```diff
@@ -396,8 +395,8 @@ def deploy_pod(self, command: Optional[List] = None):
         if self.is_pod_already_deployed():
             self.delete_pod()
 
-        pod_manifest = self.create_pod_manifest(command=command)
-        self.create_pod(pod_manifest)
+        self.set_pod_manifest(command=command)
+        self.create_pod(self.pod_manifest)
```
Comment (csomh) [inline sandcastle/api.py:399]: Is this change required? The commit message doesn't mention it or give any reasoning about it.

I think the previous version was more expressive.

File: sandcastle/api.py [hunk 2]
```diff
@@ -609,8 +608,11 @@ def exec(
                 elif status == "Failure":
                     logger.info("exec command failed")
                     logger.debug(j)
-                    logger.error(f"output = {response!r}")
-                    self._copy_mdir_from_pod(unique_dir)
+                    logger.info(f"output:\n{response}")
```
Comment (csomh) [inline sandcastle/api.py:610]: I'm not sure if it's important to consider this also from a non-packit-service POV: a user of the Sandcastle API might want to still get the content of the working directory for inspection, even if the command failed. Wouldn't be more reliable to check the status of the pod in `_copy_mdir_from_pod`? This check would also make the "Success" branch more reliable (where a similar thing could also happen: the pod is shut down after the command was successful but before a copy operation starts).

File: sandcastle/api.py (hunk 1 above)
Comment (TomasTomecek) [reply sandcastle/api.py:399]: I realized I haven't described this change precisely in the commit message after I pushed it and wanted to see CI results first, sorry about that. Though you can see how it's used in the new test case.

I need to trick sandcastle into thinking that the pod is using the default command while decreasing the timeout - so the pod manifest is a class attribute which is mutated in the test case. In this way I managed to reproduce Martin's problem in the new test case. I'm all ears if you have a better idea how to do that.

I think it makes sense to start trimming down ways how sandcastle can be invoked only to the one how we are using it in packit-service. That would enable us to drop tenths of lines of code.

File: sandcastle/api.py (hunk 2 above)
Comment (TomasTomecek) [reply sandcastle/api.py:610]: yes, that's a good point, I can change it that way, though to avoid race conditions, I think we should check if the pod is alive after we run `oc rsync`

File: sandcastle/api.py (hunk 1 above)
Comment (csomh) [reply sandcastle/api.py:399]: I see now. Probably removing the possibility to specify a custom command for `run()` could simplify things. This way only the create pod - exec workflow would be supported. But not in this PR, I guess.

Comment (TomasTomecek) [reply sandcastle/api.py:406]: agreed: https://github.com/packit/sandcastle/issues/110

File: sandcastle/api.py [hunk 3]
```diff
@@ -678,17 +693,30 @@ def _copy_path_from_pod(self, local_dir: Path, pod_dir: Path):
...
+                    f"{self.pod_name}:{pod_dir}/",  # trailing / to copy only content of dir
+                    f"{local_dir}",
+                ]
+            )
+        except Exception as ex:
+            # There is a race condition in k8s that it tells the pod is running even
+            # though it already killed an exec session and hence we couldn't copy
+            # anything from the pod
+            if not self.is_pod_running() or "not available in container" in str(ex):
+                logger.warning(
```
Comment (csomh) [inline sandcastle/api.py:713]: Shouldn't this be an 'error', given that there is an exception raised in the next line?

**Review by csomh — APPROVED** [review]
:rocket:

File: sandcastle/api.py (hunk 3 above)
Comment (TomasTomecek) [reply sandcastle/api.py:713]: the only reason I used warning was that it wouldn't pop up in sentry when this happens for timed out runs
